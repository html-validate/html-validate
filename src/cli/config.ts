module.exports = {
	extends: [
		'htmlvalidate:recommended',
	],
};
