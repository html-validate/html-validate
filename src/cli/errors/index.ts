export { handleSchemaValidationError } from "./handle-schema-validation-error";
export { ImportResolveMissingError } from "./import-resolve-missing-error";
