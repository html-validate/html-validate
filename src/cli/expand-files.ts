import fs from "fs";
import path from "node:path";
import { globSync } from "glob";

const DEFAULT_EXTENSIONS = ["html"];

/**
 * @public
 */
export interface ExpandOptions {
	/**
	 * Working directory. Defaults to `process.cwd()`.
	 */
	cwd?: string;

	/**
	 * List of extensions to search for when expanding directories. Extensions
	 * should be passed without leading dot, e.g. "html" instead of ".html".
	 */
	extensions?: string[];
}

function isDirectory(filename: string): boolean {
	const st = fs.statSync(filename);
	return st.isDirectory();
}

function join(stem: string, filename: string): string {
	if (path.isAbsolute(filename)) {
		return path.normalize(filename);
	} else {
		return path.normalize(path.join(stem, filename));
	}
}

function directoryPattern(extensions: string[]): string {
	switch (extensions.length) {
		case 0:
			return "**/*";
		case 1:
			return `**/*.${extensions[0]}`;
		default:
			return `**/*.{${extensions.join(",")}}`;
	}
}

/**
 * Takes a number of file patterns (globs) and returns array of expanded
 * filenames.
 */
export function expandFiles(patterns: string[], options: ExpandOptions): string[] {
	const cwd = options.cwd ?? process.cwd();
	const extensions = options.extensions ?? DEFAULT_EXTENSIONS;

	const files = patterns.reduce((result: string[], pattern: string) => {
		/* process - as standard input */
		if (pattern === "-") {
			result.push("/dev/stdin");
			return result;
		}

		for (const filename of globSync(pattern, { cwd })) {
			/* if file is a directory recursively expand files from it */
			const fullpath = join(cwd, filename);
			if (isDirectory(fullpath)) {
				const dir = expandFiles([directoryPattern(extensions)], { ...options, cwd: fullpath });
				result = result.concat(dir.map((cur) => join(filename, cur)));
				continue;
			}

			result.push(fullpath);
		}

		return result.sort((a, b) => {
			const pa = a.split("/").length;
			const pb = b.split("/").length;
			if (pa !== pb) {
				return pa - pb;
			} else {
				return a > b ? 1 : -1;
			}
		});
	}, []);

	/* only return unique matches */
	return Array.from(new Set(files));
}
