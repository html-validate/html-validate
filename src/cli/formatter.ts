import fs from "fs";
import path from "path";
import { Formatter } from "../formatters";
import { Report, Result } from "../reporter";

type WrappedFormatter = (results: Result[]) => string;

function wrap(
	formatter: Formatter,
	dst: string
): (results: Result[]) => string {
	return (results: Result[]) => {
		const output = formatter(results);
		if (dst) {
			const dir = path.dirname(dst);
			if (!fs.existsSync(dir)) {
				fs.mkdirSync(dir, { recursive: true });
			}
			fs.writeFileSync(dst, output, "utf-8");
			return null;
		} else {
			return output;
		}
	};
}

export function getFormatter(formatters: string): (report: Report) => string {
	const fn: WrappedFormatter[] = formatters.split(",").map((cur) => {
		const [name, dst] = cur.split("=", 2);
		const moduleName = name.replace(/[^a-z]+/g, "");
		/* eslint-disable-next-line import/no-dynamic-require */
		const formatter = require(`../formatters/${moduleName}`);
		return wrap(formatter, dst);
	});
	return (report: Report) => {
		return fn
			.map((formatter: WrappedFormatter) => formatter(report.results))
			.filter(Boolean)
			.join("\n");
	};
}
