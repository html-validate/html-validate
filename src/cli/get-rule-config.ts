import { Severity, type RuleConfig } from "..";
import { parseSeverity } from "./parse-severity";

function parseItem(value: string): { ruleId: string; severity: Severity } {
	if (value.includes(":")) {
		const [ruleId, severity] = value.split(":", 2);
		return { ruleId, severity: parseSeverity(ruleId, severity) };
	} else {
		return { ruleId: value, severity: Severity.ERROR };
	}
}

/**
 * @internal
 */
export function getRuleConfig(values: string | string[]): RuleConfig {
	if (typeof values === "string") {
		return getRuleConfig([values]);
	}

	return values.reduce<RuleConfig>((parsedRules, value) => {
		const { ruleId, severity } = parseItem(value.trim());
		return { [ruleId]: severity, ...parsedRules };
	}, {});
}
