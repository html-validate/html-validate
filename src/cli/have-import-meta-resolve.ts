/**
 * @internal
 */
export function haveImportMetaResolve(): boolean {
	return "resolve" in import.meta;
}
