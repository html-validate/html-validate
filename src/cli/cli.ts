import { readFileSync } from "fs";
import { ConfigData, UserError, HtmlValidate, Report } from "../main";
import { expandFiles, ExpandOptions } from "./expand-files";
import { getFormatter } from "./formatter";
import { IsIgnored } from "./is-ignored";
import { init, InitResult } from "./init";

const defaultConfig: ConfigData = {
	extends: ["html-validate:recommended"],
};

export interface CLIOptions {
	configFile?: string;
	rules?: string | string[];
}

export class CLI {
	private options: CLIOptions;
	private config: ConfigData;
	private ignored: IsIgnored;

	/**
	 * Create new CLI helper.
	 *
	 * Can be used to create tooling with similar properties to bundled CLI
	 * script.
	 */
	public constructor(options?: CLIOptions) {
		this.options = options || {};
		this.config = this.getConfig();
		this.ignored = new IsIgnored();
	}

	/**
	 * Returns list of files matching patterns and are not ignored.
	 */
	public expandFiles(patterns: string[], options: ExpandOptions = {}): string[] {
		return expandFiles(patterns, options).filter((filename) => !this.isIgnored(filename));
	}

	public getFormatter(formatters: string): (report: Report) => string {
		return getFormatter(formatters);
	}

	/**
	 * Initialize project with a new configuration.
	 *
	 * A new `.htmlvalidate.json` file will be placed in the path provided by
	 * `cwd`.
	 */
	public init(cwd: string): Promise<InitResult> {
		return init(cwd);
	}

	/**
	 * Searches ".htmlvalidateignore" files from filesystem and returns `true` if
	 * one of them contains a pattern matching given filename.
	 */
	public isIgnored(filename: string): boolean {
		return this.ignored.isIgnored(filename);
	}

	/**
	 * Clear cache.
	 *
	 * Previously fetched [[HtmlValidate]] instances must either be fetched again
	 * or call [[HtmlValidate.flushConfigCache]].
	 */
	public clearCache(): void {
		this.ignored.clearCache();
	}

	/**
	 * Get HtmlValidate instance with configuration based on options passed to the
	 * constructor.
	 */
	public getValidator(): HtmlValidate {
		return new HtmlValidate(this.config);
	}

	private getConfig(): ConfigData {
		const { options } = this;
		const config: ConfigData = options.configFile
			? JSON.parse(readFileSync(options.configFile, "utf-8"))
			: defaultConfig;
		let rules = options.rules;
		if (rules) {
			if (Array.isArray(rules)) {
				rules = rules.join(",");
			}
			const raw = rules
				.split(",")
				.map((x: string) => x.replace(/ *(.*):/, '"$1":'))
				.join(",");
			try {
				const rules = JSON.parse(`{${raw}}`);
				config.extends = [];
				config.rules = rules;
			} catch (e) {
				// istanbul ignore next
				throw new UserError(`Error while parsing --rule option "{${raw}}": ${e.message}.\n`);
			}
		}
		return config;
	}
}
