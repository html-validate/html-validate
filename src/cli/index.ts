import { CLI } from "./cli";
export default CLI;
