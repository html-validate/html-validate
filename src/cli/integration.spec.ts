import path from "node:path";
import { CLI } from "./cli";

const root = path.resolve(__dirname, "../../");
const cli = new CLI();
const files = cli.expandFiles(["test-files/config"]).map((it) => path.relative(root, it));
const htmlvalidate = cli.getValidator();

it.each(files)("%s", async (filename) => {
	expect.assertions(1);
	const report = await htmlvalidate.validateFile(filename);
	expect(report).toMatchSnapshot();
});
