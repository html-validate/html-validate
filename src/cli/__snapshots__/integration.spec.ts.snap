// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-files/config/cjs-config/file.html 1`] = `
Object {
  "errorCount": 0,
  "results": Array [],
  "valid": true,
  "warningCount": 0,
}
`;

exports[`test-files/config/directive/disable.html 1`] = `
Object {
  "errorCount": 2,
  "results": Array [
    Object {
      "errorCount": 2,
      "filePath": "test-files/config/directive/disable.html",
      "messages": Array [
        Object {
          "column": 3,
          "context": "i",
          "line": 1,
          "message": "<i> must not be self-closed",
          "offset": 2,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "i:nth-child(1)",
          "severity": 2,
          "size": 2,
        },
        Object {
          "column": 3,
          "context": "i",
          "line": 8,
          "message": "<i> must not be self-closed",
          "offset": 212,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "i:nth-child(4)",
          "severity": 2,
          "size": 2,
        },
      ],
      "source": "<i/>Before disable, should trigger

<!-- [html-validate-disable no-self-closing] -->
<i/>After disable, should not trigger
<i/>After disable, should not trigger

<!-- [html-validate-enable no-self-closing] -->
<i/>Before after, should trigger
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/directive/disable-block.html 1`] = `
Object {
  "errorCount": 4,
  "results": Array [
    Object {
      "errorCount": 4,
      "filePath": "test-files/config/directive/disable-block.html",
      "messages": Array [
        Object {
          "column": 3,
          "context": "i",
          "line": 1,
          "message": "<i> must not be self-closed",
          "offset": 2,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "i:nth-child(1)",
          "severity": 2,
          "size": 2,
        },
        Object {
          "column": 4,
          "context": "i",
          "line": 3,
          "message": "<i> must not be self-closed",
          "offset": 43,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "div:nth-child(2) > i:nth-child(1)",
          "severity": 2,
          "size": 2,
        },
        Object {
          "column": 3,
          "context": "i",
          "line": 11,
          "message": "<i> must not be self-closed",
          "offset": 335,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "i:nth-child(3)",
          "severity": 2,
          "size": 2,
        },
        Object {
          "column": 3,
          "context": "i",
          "line": 16,
          "message": "<i> must not be self-closed",
          "offset": 439,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "i:nth-child(5)",
          "severity": 2,
          "size": 2,
        },
      ],
      "source": "<i/>Outside block, should trigger
<div>
	<i/>Inside block, before directive, should trigger
	<!-- [html-validate-disable-block no-self-closing] -->
	<i/>Inside block, after directive, shout not trigger
	<div>
		<i/>Inside block, after directive, shout not trigger
	</div>
	<i/>Inside block, after directive, shout not trigger
</div>
<i/>Outside block, should trigger

<div>
	<!-- [html-validate-disable-block no-self-closing] -->
</div>
<i/>Outside block, should trigger

<!-- [html-validate-disable-block no-self-closing] -->
Should handle when root element is parent but no children
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/directive/disable-multiple.html 1`] = `
Object {
  "errorCount": 2,
  "results": Array [
    Object {
      "errorCount": 2,
      "filePath": "test-files/config/directive/disable-multiple.html",
      "messages": Array [
        Object {
          "column": 2,
          "context": Object {
            "documentation": "\`<blink>\` has no direct replacement and blinking text is frowned upon by accessibility standards.",
            "source": "non-standard",
            "tagName": "blink",
          },
          "line": 1,
          "message": "<blink> is deprecated",
          "offset": 1,
          "ruleId": "deprecated",
          "ruleUrl": "https://html-validate.org/rules/deprecated.html",
          "selector": "blink",
          "severity": 2,
          "size": 5,
        },
        Object {
          "column": 7,
          "context": "blink",
          "line": 1,
          "message": "<blink> must not be self-closed",
          "offset": 6,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "blink",
          "severity": 2,
          "size": 2,
        },
      ],
      "source": "<blink/> Outside block, should trigger

<div>
	<!-- [html-validate-disable-block no-self-closing, deprecated] -->
	<blink/>Inside block, should not trigger
</div>
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/directive/disable-next.html 1`] = `
Object {
  "errorCount": 2,
  "results": Array [
    Object {
      "errorCount": 2,
      "filePath": "test-files/config/directive/disable-next.html",
      "messages": Array [
        Object {
          "column": 3,
          "context": "i",
          "line": 1,
          "message": "<i> must not be self-closed",
          "offset": 2,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "i:nth-child(1)",
          "severity": 2,
          "size": 2,
        },
        Object {
          "column": 3,
          "context": "i",
          "line": 5,
          "message": "<i> must not be self-closed",
          "offset": 136,
          "ruleId": "no-self-closing",
          "ruleUrl": "https://html-validate.org/rules/no-self-closing.html",
          "selector": "i:nth-child(3)",
          "severity": 2,
          "size": 2,
        },
      ],
      "source": "<i/>Before disable, should trigger

<!-- [html-validate-disable-next no-self-closing] -->
<i/>First after disable, should not trigger
<i/>Second after disable, should trigger
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/elements/file.html 1`] = `
Object {
  "errorCount": 2,
  "results": Array [
    Object {
      "errorCount": 2,
      "filePath": "test-files/config/elements/file.html",
      "messages": Array [
        Object {
          "column": 3,
          "context": Object {
            "tagName": "deprecated",
          },
          "line": 2,
          "message": "<deprecated> is deprecated",
          "offset": 8,
          "ruleId": "deprecated",
          "ruleUrl": "https://html-validate.org/rules/deprecated.html",
          "selector": "div > deprecated",
          "severity": 2,
          "size": 10,
        },
        Object {
          "column": 4,
          "context": undefined,
          "line": 5,
          "message": "Element <li> is not permitted as content in <ul>",
          "offset": 67,
          "ruleId": "element-permitted-content",
          "ruleUrl": "https://html-validate.org/rules/element-permitted-content.html",
          "selector": "div > ul > li",
          "severity": 2,
          "size": 2,
        },
      ],
      "source": "<div>
	<deprecated>new (deprecated) element</deprecated>

	<ul>
		<li>overridden so it is disallowed</li>
		<p>overridden so it is allowed</p>
	</ul>
</div>
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/extend-order/foo.html 1`] = `
Object {
  "errorCount": 1,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": "test-files/config/extend-order/foo.html",
      "messages": Array [
        Object {
          "column": 3,
          "context": undefined,
          "line": 3,
          "message": "Element <a> is not permitted as content in <div>",
          "offset": 94,
          "ruleId": "element-permitted-content",
          "ruleUrl": "https://html-validate.org/rules/element-permitted-content.html",
          "selector": "div:nth-child(1) > a",
          "severity": 2,
          "size": 1,
        },
      ],
      "source": "<div>
	<!-- should yield error: allowed in default.json but disallowed by override.json -->
	<a>..</a>
</div>

<div>
	<!-- should not yield error -->
	<b>..</b>
</div>
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/extend-order/nested/foo.html 1`] = `
Object {
  "errorCount": 2,
  "results": Array [
    Object {
      "errorCount": 2,
      "filePath": "test-files/config/extend-order/nested/foo.html",
      "messages": Array [
        Object {
          "column": 3,
          "context": undefined,
          "line": 3,
          "message": "Element <a> is not permitted as content in <div>",
          "offset": 70,
          "ruleId": "element-permitted-content",
          "ruleUrl": "https://html-validate.org/rules/element-permitted-content.html",
          "selector": "div:nth-child(1) > a",
          "severity": 2,
          "size": 1,
        },
        Object {
          "column": 3,
          "context": undefined,
          "line": 4,
          "message": "Element <b> is not permitted as content in <div>",
          "offset": 81,
          "ruleId": "element-permitted-content",
          "ruleUrl": "https://html-validate.org/rules/element-permitted-content.html",
          "selector": "div:nth-child(1) > b",
          "severity": 2,
          "size": 1,
        },
      ],
      "source": "<div>
	<!-- should both yield errors: disallowed by nested.json -->
	<a>..</a>
	<b>..</b>
</div>

<div>
	<!-- should not yield error -->
	<c>..</d>
</div>
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/file.html 1`] = `
Object {
  "errorCount": 1,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": "test-files/config/file.html",
      "messages": Array [
        Object {
          "column": 6,
          "context": "br",
          "line": 2,
          "message": "End tag for <br> must be omitted",
          "offset": 49,
          "ruleId": "void-content",
          "ruleUrl": "https://html-validate.org/rules/void-content.html",
          "selector": null,
          "severity": 2,
          "size": 3,
        },
      ],
      "source": "<!-- default configuration yields error -->
<br></br>
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/js-config/file.html 1`] = `
Object {
  "errorCount": 0,
  "results": Array [],
  "valid": true,
  "warningCount": 0,
}
`;

exports[`test-files/config/off/error/file.html 1`] = `
Object {
  "errorCount": 1,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": "test-files/config/off/error/file.html",
      "messages": Array [
        Object {
          "column": 6,
          "context": "br",
          "line": 2,
          "message": "End tag for <br> must be omitted",
          "offset": 42,
          "ruleId": "void-content",
          "ruleUrl": "https://html-validate.org/rules/void-content.html",
          "selector": null,
          "severity": 2,
          "size": 3,
        },
      ],
      "source": "<!-- reconfigured to yield error -->
<br></br>
",
      "warningCount": 0,
    },
  ],
  "valid": false,
  "warningCount": 0,
}
`;

exports[`test-files/config/off/file.html 1`] = `
Object {
  "errorCount": 0,
  "results": Array [],
  "valid": true,
  "warningCount": 0,
}
`;

exports[`test-files/config/warn/file.html 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "test-files/config/warn/file.html",
      "messages": Array [
        Object {
          "column": 6,
          "context": "br",
          "line": 2,
          "message": "End tag for <br> must be omitted",
          "offset": 44,
          "ruleId": "void-content",
          "ruleUrl": "https://html-validate.org/rules/void-content.html",
          "selector": null,
          "severity": 1,
          "size": 3,
        },
      ],
      "source": "<!-- configured to yield a warning -->
<br></br>
",
      "warningCount": 1,
    },
  ],
  "valid": true,
  "warningCount": 1,
}
`;
